--- 
title: "Differentially expressed gene analysis of RNA-seq data using R"
site: bookdown::bookdown_site
output: bookdown::gitbook
documentclass: book
classoption: a4paper
---

```{r, echo=FALSE, include=FALSE}
library(knitr)
knitr::opts_chunk$set(cache=TRUE)
```

# NGS data exploration

When working on RNA-Seq data using R the short reads are usually mapped to the 
reference genome using a genome mapper and the DNA sequences themselves are 
handled by the mapping software. However, it is useful to know that we can perform 
a number of analyses on the DNA sequences themselves in R. The QC analysis and 
evaluation of the sequence collection content can be managed as well.

There are a number of different DNA sequencing platforms in use that have their 
own characteristics and challenges.

## Required packages for these exercises

If you would like to run these exercises again on your own machine, you will
need to install the following R packages.

* `mixOmics`
* `tidyr`
* `ggplot2`

```{r, eval=F}
install.packages(c('mixOmics','tidyr','ggplot2'))
```

Bioconductor packages:

* `Biostrings`
* `ShortRead`
* `Rsamtools`
* `Rsubread`
* `edgeR`
* `limma`
* `org.Hs.eg.db`


<!--chapter:end:index.Rmd-->

## Reading from a fasta file

FASTA format is a text-based format for representing nucleotide and peptide 
sequences using their single-letter IUPAC codes. The format also allows for 
sequence names and comments to precede the sequences. The format originates from 
the FASTA software package, but has now become a standard in the field of bioinformatics.

The simplicity of FASTA format makes it easy to manipulate and parse sequences 
using text-processing tools that are built into R. A number of packages 
make the process of loading a fasta file very much easier.

```{r read-fasta}
library(Biostrings)
library(ShortRead)

NGS_DIR <- "../data/NGS_sequences"
comt <- readFasta(file.path(NGS_DIR, "comt.fasta"))
print(comt)
```


```{r fasta-id}
id(comt)
```

```{r fasta-sread}
sread(comt)
```

```{r fasta-stat}
width(comt)
length(comt)
```

From this code it can be seen that we have created an object of `ShortRead` type 
that contains a `DNAStringSet` containing two DNA sequences. The `width()` 
function reports the length of each DNA sequence while the `length()` function 
reports the number of DNA sequences in the sequence collection.

The character representation of the sequence remains accessible:

```{r fasta-seq}
comtStr <- toString(sread(comt[1]))
class(comtStr)
substring(comtStr,1,50)
```

Find all positions in the sequence with "ATG" codon:
```{r fasta-grep}
gregexpr("ATG", comtStr)
as.vector(gregexpr("ATG", comtStr)[[1]])
```


There are a wide range of different functions that can be applied to the manipulation
of individual DNA sequences. We could apply functions that include `strsplit`, 
`replaceLetterAt`, `subseq`, `maskMotif`, `reverseComplement` etc. These methods 
work well but for single sequences or for a small collection of sequences, for
batch jobs, other software might be more suitable.

There are a number of other convenience utilities in R/Bioconductor:

```{r genetic-code}
GENETIC_CODE
```

```{r iupac-map}
IUPAC_CODE_MAP
```

## Parsing FastQ files

For a more comprehensive review of what can be performed using `R/Bioconductor` 
for short reads let us have a look at some data from the public domain. The 
Short Read Archive (SRA) hosted by the NBCI and provides a public repository 
where high dimensional data can be shared with the community. Study `ERX337002` 
looks at the metagenomics of food (http://www.ncbi.nlm.nih.gov/sra/ERX337002).
The Ion Torrent platform has been used to sequence the microbial content of an artisanal cheese. 

The raw data has been downloaded from the public domain and has been converted 
from the `sra` format into a `fastq` file format (using SRA Toolkit) that can be 
used with a number of packages. The fastq file prepared has been gzipped to 
save space (and increase read performance).
  

```
curl -0 ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/ERX/ERX337/ERX337002/
ERR364233.sra
fastq-dump ERR364233.sra
gzip ERR364233.fastq
```

A brief summary of the content in the file can be prepared by loading the fastq 
file with the `readFastq` function.

```{r readfastq}
cheese <- readFastq(file.path(NGS_DIR, "ERR364233.subset.fastq.gz"))
print(cheese)
```

```{r fastqsummary}
qaSummary <- qa(cheese, NGS_DIR)
report(qaSummary, dest="./results")
```

```{block, type="rmdexercise"}
1. Enter the above commands.
2. Click on the **Files** tab in the bottom-right corner of RStudio, then click on **More** > **Go to working directory**.
3. Click on **New Folder** and create the `results` folder.
4. Run the `qaSummary()` and `report()` commands as above, if not already done.
5. Click on the **results** folder and the **index.html** file.
6. Select the **View in Web Browser** (if a popup window appears click Try again).
7. Have a look a the report from the `qaSummary()` function.
```


This shows that there are 200K sequence reads in the sequence collection 
that have a length of between 8 and 358 nucleotides. Let's have a look at the distribution of read lengths:

```{r read-histogram}
hist(width(cheese), breaks=100)
```

We recommend that we do not use such short reads in our sequence collection.
Filter out sequences that are less than 100 nt in length and replot the distribution.

```{r read-filtering}
cheese.filtered <- cheese[-which(width(cheese) < 100)]
hist(width(cheese.filtered), breaks=100)
```

```{block, type="rmdexercise"}
How many reads were removed from the filtering step? _________
(**hint** `length()`)
```

We have looked at the distribution of sequence lengths in the collection. We can
also check the distribution of individual nucleotides in the collection. 
The `alphabetFrequency` function collects frequency counts for each of the 
IUPAC nucleotides in a `DNAStringSet` object. The method will generate a table 
that be summarised and plotted using your preferred graphics library.

```{r read-baseFreq}
freq <- alphabetFrequency(sread(cheese.filtered))
totalCount <- colSums(freq)
barplot(totalCount[1:4])
```

```{block, type="rmdexercise"}
1. Run the command above.
2. What *data type* is the variable `freq`? ________
3. What are the dimensions of `freq`? ______________
4. Why did we only plot the first 4 elements of `totalCount`? ________


```

Similar plots can be prepared for di- and trinucleotide frequencies, this looks 
at all possible words. You can similarly look for hexanucleotide frequencies 
or any other number, check the `alphabetFrequency()` documentation. This can
assists in detecting any over-represented subsequences.

```{r read-dinucFreq}
barplot(colSums(dinucleotideFrequency(sread(cheese.filtered))),cex.axis = 0.75, las=2)
```


Depending on the sequencing strategy that you applied, you may have ligated 
synthetic adapter sequences, primers and other constructs within the target 
amplicons. There are a number of methods for stripping such synthetic sequences 
from your sequence collection.

```{r read-trimming}
head(sread(cheese.filtered))
cloningPrimer <- "AACATTTGTCCTAATCATTGGGGATA"
cheese.clipped <- trimLRPatterns(Lpattern = cloningPrimer, subject=cheese.filtered,
                                 max.Lmismatch = 0.33)
head(sread(cheese.clipped))
```

There is of course a lot more that can be done to allow for perfect, imperfect and 
truncated primer sequences or adapters at either the forward or reverse ends of the sequence.

One of the reasons why the `fastq` file format has become so popular in genomics is 
that in addition to storing information relating to the DNA sequence, the file stores
information on the DNA sequence and the quality (confidence score that the base has been measured correctly). 

The quality scores in a ShortRead data collection are accessible using the `quality()` function.

```{r read-quality-compare}
head(quality(cheese.clipped))
encoding(quality(cheese.clipped))
cheese.quality <- as(quality(cheese.clipped), "matrix")
hist(rowMeans(cheese.quality[sample(nrow(cheese.quality), 25000),],na.rm=TRUE), 
     main="", xlab="")
```

```{block, type="rmdexercise"}
1. Type in the commands above.
2. What *data type* is the variable `cheese.quality`? ______________
3. What is the dimension of `cheese.quality`? ______________________
4. How does the number of columns relate to the length of the reads? _____
  
5. Below are the broken-down parts of the `hist(...)` function that produced
   the plot above. The list starts with the most inner function and works it's
   way out. Can you describe what each inner function is doing and the expected result?
   The letter in the curly braces e.g. \{a\} means substitute the result
   from the part \{a\} into this function.
   
   a. `nrow(cheese.quality)`: ______________________
   b. `sample({a}, 25000)`  : ______________________
   c. `cheese.quality[{b},]` : _____________________
   d. `rowMeans({c}, na.rm=TRUE)`: _________________
   e. `hist({d}, main="", xlab="")`: _______________
```

The code above presents a text format (representation of the integer based quality scores) 
of the quality data. This is converted into a qualityMatrix that can then be
plotted to show the distribution of quality scores across the sequence collection.

Plotting the per-base quality scores is the goal of many software applications 
and this can be managed simply in R as well. One of the requirements to produce 
these plots is however, to reduce the dimensions of the data. Plotting the 
characteristics for every read is unnecessary since a sample should convey 
trends within the data.

```{r read-boxplot-quality}
boxplot(cheese.quality[sample(nrow(cheese.quality), 25000), 1:100],out.cex=0.5)
```


```{r read-boxplot-quality2}
mung <- function(i) {
  lower <- i
  upper <- i + 4
  if (upper > ncol(cheese.quality)) {
    upper <- ncol(cheese.quality)
  }
  return(rowMeans(cheese.quality[,seq(lower,upper)], na.rm=TRUE))
}

sequence <- seq(1, ncol(cheese.quality), 5)
groupedQuality <- as.data.frame(sapply(sequence, mung))
colnames(groupedQuality) <- as.character(sequence)

boxplot(groupedQuality[sample(nrow(groupedQuality), 25000),], outcex=0.5)
abline(h=c(20,28),lty=2,col='white')
rect(-5,28,358,45,col=rgb(0.1,1,0,alpha=0.25),border=NA)
rect(-5,20,358,28,col=rgb(1,1,0,alpha=0.25),border=NA)
rect(-5,0,358,20,col=rgb(1,0.6,0.6,alpha=0.25),border=NA)
```

Having prepared this box-and-whisker plot we can see that the distribution of 
quality scores is generally OK but there are quite a large number of low quality 
bases described as the interquartile ranges within the whiskers. To make the 
data most valuable to the analysis it would be worthwhile to filter the sequences 
to exclude the substrings of lowest quality from the 3'-end forwards. The 
`ShortRead` package implements some easy to use functions such as `trimTails` and 
`trimTailw`.

```{r read-boxplot-quality3}
cheese.clipped <- trimTailw(cheese.clipped, k=4, a="4", halfwidth=5)
cheese.quality <- as(quality(cheese.clipped), "matrix")
dim(cheese.quality)
sequence <- seq(1, ncol(cheese.quality), 5)
groupedQuality <- as.data.frame(sapply(sequence, mung))
colnames(groupedQuality) <- as.character(sequence)
boxplot(groupedQuality[sample(nrow(groupedQuality), 25000),], outcex=0.5)
abline(h=c(20,28),lty=2,col='white')
rect(-5,28,358,45,col=rgb(0.1,1,0,alpha=0.25),border=NA)
rect(-5,20,358,28,col=rgb(1,0.8,0,alpha=0.25),border=NA)
rect(-5,0,358,20,col=rgb(1,0.6,0.6,alpha=0.25),border=NA)
```

This shows that we have greatly improved the overall structure of the data and 
have removed a large number of the less-than-perfect bases. This example may 
have been a little more aggressive than we would really wish to apply in the laboratory setting.

The `ShortRead` package has a well implemented framework for filtering 
sequences and it would be simple to implement other filters as required. The 
filter can be illustrated by filtering out the homopolymers from the sequence 
collection. These provide mechanisms for restricting the sequence collection 
on the basis of e.g. base composition, number of N-residues and even relative abundance.

A whole lot more can be done using the `ShortRead` and `Biostrings` packages. 

```{block, type="rmdexercise"}
**Prepare some QC information for an Illumina data collection**

Using the RNA-Seq reads that you will perform differential expression with,
prepare a brief report of the data. Using the `qaSummary` function to prepare
a synopsis of the library content and investigate the per-base quality
scores across the sequence collection. Would you recommend that we re-run
the analysis after some data trimming?

Have a look at the over-represented k-mer words in the sequence collection 
and see if you can create a filter to strip-out the sequence that contain the
most abundant k-mer.
```

<!--chapter:end:01-qqch.Rmd-->

---
title: "Differentially expressed protein analysis of SWATH data using ANOVA testing"
output: html_notebook
---

This worflow shows how to process data from SWATH-MS protein quantification, in order to study differential expression of the proteins between the different conditions.

The data in this example come from : Bjelosevic S, Pascovici D, Ping H, et.al. Quantitative age-specific variability of plasma proteins in healthy neonates, children and adults. Molecular & Cellular Proteomics (March 23 2017). Four conditions are compared each conditions have ten replicates :  
-neonates  
-child less than one year old  
-child between 1 and five years old  
-adults  

146 proteins were identified and quantified. 

Here we use ANOVA (analysis of variance) to identify significantly differentially expressed protein. The ANOVA test the statistical significance of the mean difference between several groups. But ANOVA doesn't allow to identify between exactly which groups this difference is significant that is why a POST HOC test is needed to precise the groups between which the difference is the most significant.

#1. loading and visualizing the data 
```{r}
data<-read.table("SWATH.example.data.csv", sep=",",header=T, row.names = 1)
data
boxplot(data, ylab="Intensity", xlab="samples", main = "distribution of Intensity") 
hist(as.numeric(unlist(data)), main = "Histogram of Intensity distribution",  xlab = "Intensity")
```

#2. log transform
The data is log transformed so that:  
-the spread is even across the intensity range  
-the variability stays constant at all intensity levels  
-the distribution of experimental errors is normal  
-the distribution of intensity is bell shaped  

```{r}
data <- log2(data)
boxplot(data, ylab="Intensity", xlab="samples") 
hist(as.numeric(unlist(data)), main = "Histogram of Intensity distribution",  xlab = "Intensity")
```


#3. Normalization

Normalization allows to resolve the bias and errors introduced between the samples by the experimental tools or the samples preparation. Normalization relies on the fact that most of the proteins are expected to be expressed the same way between the samples, so it is necessary to remove the differences induced by the experiment to not overestimate the differencial expression. here we use the mean center and scaling normalization : centering assure that the mean of the different samples is equal (set to 0), and scaling sets the standard deviation to 1.  
```{r}
norm.data <- scale(data) ##mean center and scaling normalization
boxplot(norm.data, ylab="Intensity", xlab="samples")
```


#4. Differential expression annalysis
##4.1 ANOVA test

The type of ANOVA test to perform is a one-way ANOVA on each protein.
```{r}
Results <- as.data.frame(norm.data)
## Calcul of the mean Intensity for each conditions : 
Results$control <- rowMeans(norm.data[, c(1:10)])
Results$less1 <- rowMeans(norm.data[, c(11:20)])
Results$onetofive <- rowMeans(norm.data[, c(21:30)])
Results$adult <- rowMeans(norm.data[, c(31:40)])
Results <- Results[, c(41, 42, 43, 44)]


d <- data.frame(mean = 1:40, group = 1:40)
for (i in 1:nrow(norm.data)) {
  d$mean <- as.numeric(norm.data[i, ])
  d$group[1:10] <- "control"
  d$group[11:20] <- "less1"
  d$group[21:30] <- "onetofive"
  d$group[31:40] <- "adult"
  model <- lm(formula = mean ~ group, data = d)
  Results$p[i] <- anova(model)$"Pr(>F)"
  Results$p[i] <- p.adjust(Results$p[i], method = "BH") ## multiple testing correction with de Benjamini Hochberg technique
  # print(TukeyHSD(aov(mean ~ group, data = d)))
}
```

##4.3 Differentially expressed proteins


Calcul of the maximum fold change for each protein (difference between the highest and the lowest intensity mean) : 
```{r}
Results$FC <-apply(Results[, 1:4], 1, max) - apply(Results[, 1:4], 1, min)
Results
```


A protein is considered differentially expressed if its p.value and fold change are superior to a determined threshold. The fold change represent how much a protein is differentially expressed between two conditions and the p value allows to evaluate the statistical significance of this difference.
```{r}
resultsANOVA <-Results[which(Results$p <= 0.05 & abs(Results$FC) >= 0.5), ]
resultsANOVA
```


##4.4 post hoc analysis

The post hoc analysis is done using the tukey's test to determine the significants pairwise comparisons.
```{r}
DEprot <- as.data.frame(data[which(row.names(data) %in% row.names(resultsANOVA)),])
d <- data.frame(mean = 1:40, group = 1:40)
for (i in 1:nrow(DEprot)) {
  d$mean <- as.numeric(DEprot[i, ])
  d$group[1:10] <- "control"
  d$group[11:20] <- "less1"
  d$group[21:30] <- "onetofive"
  d$group[31:40] <- "adult"
  print(paste0(row.names(DEprot[i,]), " :"))
  print(TukeyHSD(aov(mean ~ group, data = d)))
}
```





```{r}
heat <- as.matrix(data[which(row.names(data) %in% row.names(resultsANOVA)),])
```



```{r}
library(gplots)
heatmap.2(heat, trace = "none") 
heatmap.2(as.matrix(data), trace = "none") 
```


```{r}
library(reshape2)
heat <- melt(heat)
names(heat) <- c("protein", "sample", "value")

library(ggplot2)
ggplot(heat, aes(sample, protein )) +
  geom_tile(aes(fill = value), color = "white") +
  scale_fill_gradient(low = "red", high = "steelblue") +
  ylab("List of genes ") +
  xlab("List of patients") +
  theme(legend.title = element_text(size = 10),
        legend.text = element_text(size = 12),
        plot.title = element_text(size=16),
        axis.title=element_text(size=14,face="bold"),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  labs(fill = "Expression level")
```


```{r}
  plotTitle <- "Results of the ANOVA tests perform on each protein"
  values <- as.data.frame(Results)
  forplot <- data.frame(x=as.numeric(values[,6]), y=-log10(values[,5]), id=row.names(values))
  tmp <- forplot[as.numeric(forplot$y)>=-log10(tresh_fc) & abs(forplot$x)>tresh_fc,]
  p <- ggplot(forplot) + geom_point(aes(x, y , color = ifelse(y>=-log10(tresh_p) & abs(x)>=tresh_fc, "not signi", "FC")),show.legend = F) +
    scale_color_manual(values = c("blue", "red")) +
    geom_text_repel(data = subset(forplot, abs(forplot$x)>=tresh_fc & forplot$y>=-log10(tresh_p)),
                    aes(x,y,label = id),
                    size = 2) +
    geom_vline(xintercept = tresh_fc ) +
    geom_vline(xintercept = -tresh_fc) + 
    geom_hline(yintercept = -log10(tresh_p)) + 
    labs(title = plotTitle,x="maximum Fold-change", y="-log10(P.Value)") + theme_bw() 
    print(p)
```


<!--chapter:end:SWATH-anova.Rmd-->

---
title: "Comparison of statistical test for SWATH differential expression analysis"
output: html_notebook
---

<!--chapter:end:SWATH-comparison.Rmd-->

---
title: "Differentially expressed protein analysis of SWATH data using the LIMMA package"
output: html_notebook
---

This worflow shows how to process data from SWATH-MS protein quantification, in order to study differential expression of the proteins between the different conditions.

The data in this example come from : Bjelosevic S, Pascovici D, Ping H, et.al. Quantitative age-specific variability of plasma proteins in healthy neonates, children and adults. Molecular & Cellular Proteomics (March 23 2017). Four conditions are compared each conditions have ten replicates :  
-neonates  
-child less than one year old  
-child between 1 and five years old  
-adults  

146 proteins were identified and quantified. 

Here we gonna use the limma package to perform the normalization of the data and the differential expression annalysis.
Limma is a Bioconductor package which uses linear model for analysing experiments and the assessment of differential expression.

#1. loading and visualizing the data 
```{r}
data<-read.table("SWATH.example.data.csv", sep=",",header=T, row.names = 1)
data
boxplot(data, ylab="log2(Intensity)", xlab="samples", main = "distribution of Intensity") 
hist(as.numeric(unlist(data)), main = "Histogram of Intensity distribution",  xlab = "Intensity")
```

#3. Normalization
##3.1 design and contrast matrix

In order to perform the normalization with the limma package we need to define the design matrix generated with model.matrix(), which identify which sample belong to which condition.    


```{r}
#for this specific data :
exp.design <- data.frame(samples = colnames(data), condition = 1) ## first we define a matrix containing the experiment design information by associating a sample name with the condition name
exp.design$condition[1:10] = "control"
exp.design$condition[11:20] = "lessone"
exp.design$condition[21:30] = "onetofive"
exp.design$condition[31:40] = "adult"

design <- model.matrix(~0 + exp.design$condition, data = exp.design) ## model.matrix() use the experiment design to generate a matrix.
colnames(design) <- sort(unique(exp.design$condition))
row.names(design) <- exp.design$samples
as.data.frame(design)
```


We also need to create a contrast matrix generated with the limma function makeContrast(). This matrix allows to define the comparisons between the differents conditions to perform.

```{r}
contrast <- makeContrasts(adult-control, lessone-control, onetofive-control, levels=design)
as.data.frame(contrast)
```


##3.2 applying limma

Now, we can normalise the dataset using the following commands. The calcNormFactors(), calculates the normalization factors to scale the library sizes.

The limma package (since version 3.16.0) offers the voom function that will normalise read counts and apply a linear model to the normalised data before computing moderated t-statistics of differential expression.

The returned data object consists of a few attributes, which you can check using names(y), one of which is the normalised expression (y$E) values in log2 scale.

```{r}
library(limma)
library(edgeR)

dge <- DGEList(data)
dge <- calcNormFactors(dge)
y <- voom(dge, design)
norm.data <- y$E
as.data.frame(norm.data)
boxplot(norm.data, ylab="normalized Intensity", xlab="samples", main = "distribution of the normalized Intensity")
hist(norm.data, main = "Histogram of Intensity distribution",  xlab = "Intensity")
```


#4. Differential expression annalysis
##4.1 fitting the model

To fit the model, use the lmFit() function, which takes in the normalised data object and the design matrix:

```{r}
fit <- lmFit(y, design)
```

Refit the model using the comparisons defined:
```{r}
fit2 <- contrasts.fit(fit, contrast)
fit2 <- eBayes(fit2)
```


##4.2 Extracting the results 

The topTable function summarises the output from limma in a table format.
```{r}
adult.vs.c <- topTable(fit2, coef = "adult - control", n = nrow(fit2))
lessone.vs.control <- topTable(fit2, coef = "lessone - control", n = nrow(fit2))
onetofive.vs.control <- topTable(fit2, coef = "onetofive - control", n = nrow(fit2))
adult.vs.c ; lessone.vs.control ; onetofive.vs.control
```


We keep only the values of interest like the p value, the adjusted p value and the fold change to create the volcano plot. The results for the different pairwise comparison are stored in a list of dataframe. 
```{r}
adult.vs.c$protein <- row.names(adult.vs.c)
row.names(adult.vs.c) <- NULL
adult.vs.c <- adult.vs.c[, c(7, 4, 5, 1)]

lessone.vs.control$protein <- row.names(lessone.vs.control)
row.names(lessone.vs.control) <- NULL
lessone.vs.control <- lessone.vs.control[, c(7, 4, 5, 1)]

onetofive.vs.control$protein <- row.names(onetofive.vs.control)
row.names(onetofive.vs.control) <- NULL
onetofive.vs.control <- onetofive.vs.control[, c(7, 4, 5, 1)]

Results <- list(adult.vs.c, lessone.vs.control, onetofive.vs.control)
```




##4.2 Differentially expressed proteins

A protein is considered differentially expressed if its p.value and fold change are superior to a determined threshold. The fold change represent how much a protein is differentially expressed between two conditions and the p value allows to evaluate the statistical significance of this difference.

```{r}

thresh_fc <- 0.5
thresh_p <- 0.05
deProt <- Results
for(i in 1:length(Results)){
    fc = as.data.frame(deProt[[i]])[,4]
    p = as.data.frame(deProt[[i]])[,3]
    dt <-as.data.frame(deProt[[i]])
    deProt[[i]] <- dt[which(p<=thresh_p & abs(fc)>=thresh_fc),]
    print(as.data.frame(deProt[[i]]))
  }
```




##4.3 Volcano Plot

For each pairwise comparison in the results list a volcano plot is created, volcano plots are the best way to visualise the p value and the fold change at the same time. 

```{r}

library(ggplot2)
library(ggrepel)

tresh_fc = 0.5 ## Fold change threshold
tresh_p = 0.05 ## p.value treshold

for(i in 1:length(Results)){
  
  plotTitle <- substr(colnames(Results[[i]])[2], 9 ,nchar(colnames(Results[[i]])[2]))
  values <- as.data.frame(Results[[i]])
  forplot <- data.frame(x=as.numeric(values[,4]), y=-log10(values[,3]), id=as.character(values[,1]))
  tmp <- forplot[as.numeric(forplot$y)>=-log10(tresh_fc) & abs(forplot$x)>tresh_fc,]
  p <- ggplot(forplot) + geom_point(aes(x, y , color = ifelse(y>=-log10(tresh_p) & abs(x)>=tresh_fc, "not signi", "FC")),show.legend = F) +
    scale_color_manual(values = c("blue", "red")) +
    geom_text_repel(data = subset(forplot, abs(forplot$x)>=tresh_fc & forplot$y>=-log10(tresh_p)),
                    aes(x,y,label = id),
                    size = 2) +
    geom_vline(xintercept = tresh_fc ) +
    geom_vline(xintercept = -tresh_fc) + 
    geom_hline(yintercept = -log10(tresh_p)) + 
    labs(title = plotTitle,x="log2(Fold-change)", y="-log10(P.Value)") + theme_bw() 
    print(p)
}
```


```{r}
boxplot(apply(norm.data,1,sd),apply(scale(log2(data)),1,sd), log(apply(data,1,sd)),main = "comparison of the standard deviation \n of the differents normalization techniques", xlab="technique", ylab="standard deviation",names=c("voom", "center and scaling", "original data"))
```



<!--chapter:end:SWATH-limma.Rmd-->

---
title: "Differentially expressed protein analysis of SWATH data using student's test"
output: html_notebook
---

This worflow shows how to process data from SWATH-MS protein quantification, in order to study differential expression of the proteins between the different conditions.

The data in this example come from : Bjelosevic S, Pascovici D, Ping H, et.al. Quantitative age-specific variability of plasma proteins in healthy neonates, children and adults. Molecular & Cellular Proteomics (March 23 2017). Four conditions are compared each conditions have ten replicates :  
-neonates  
-child less than one year old  
-child between 1 and five years old  
-adults  

146 proteins were identified and quantified. 

#1. loading and visualizing the data 
```{r}
data<-read.table("SWATH.example.data.csv", sep=",",header=T, row.names = 1)
data
boxplot(data, ylab="Intensity", xlab="samples") 
hist(as.numeric(unlist(data)), main = "Histogram of Intensity distribution",  xlab = "Intensity")
```

#2. log transform
The data is log transformed so that:  
-the spread is even across the intensity range  
-the variability stays constant at all intensity levels  
-the distribution of experimental errors is normal (necessary to apply most of the statistical test)   
-the distribution of intensity is bell shaped  

```{r}
data <- log2(data)
boxplot(data, ylab="Intensity", xlab="samples") 
hist(as.numeric(unlist(data)), main = "Histogram of Intensity distribution",  xlab = "Intensity")
```

#3. Normalization
Normalization allows to resolve the bias and errors introduced between the samples by the experimental tools or the samples preparation. Normalization relies on the fact that most of the proteins are expected to be expressed the same way between the samples, so it is necessary to remove the differences induced by the experiment to not overestimate the differencial expression. here we use the mean center and scaling normalization : centering assure that the mean of the different samples is equal (set to 0), and scaling sets the standard deviation to 1.  
  

```{r}
norm.data <- scale(data) ##mean center and scaling normalization
boxplot(norm.data, ylab="Intensity", xlab="samples")

```


#4. Differential expression annalysis
##4.1 experimental design

In order to perform the pairwise comparison we need to define the experimental design :  
- the design matrix generated with model.matrix(), which identify which sample belong to which condition. 
```{r}
#for this specific data :
exp.design <- data.frame(samples = colnames(data), condition = 1)
exp.design$condition[1:10] = "control"
exp.design$condition[11:20] = "lessone"
exp.design$condition[21:30] = "onetofive"
exp.design$condition[31:40] = "adult"

design <- model.matrix(~0 + exp.design$condition, data = exp.design)
colnames(design) <- sort(unique(exp.design$condition))
row.names(design) <- exp.design$samples
as.data.frame(design)
```



- the contrast matrix generated with makeContrasts(), which define the comparison to perform between the differents conditions.
```{r}
library(limma)
contrast <- makeContrasts(adult-control, lessone-control, onetofive-control, adult+lessone+onetofive-control, levels=design)
as.data.frame(contrast)
```



##4.2 Fold change and statistical test calcul

For each comparison and for each protein the fold change and the student p value are calculated. the results are stocked in a list of data frame (each data frame is the result for one comparison defined nin the contrast matrix).
The fold change is the mean intensity difference between two condition, a fold change of zero means no difference a negative one a down-regulation and a positive one an up-regulation.

```{r}
  ## in this example there are four pairwise comparison to perform:  


  nbComp<-ncol(contrast)
  Results<-list() ## a list of data frames containing the results
  
  for (i in 1:nbComp){ ## for each comparison defined in the contrast matrix :
    
    c1<-row.names(contrast)[contrast[,i] == 1 ] ## names of the first conditions of the pairwise comparison
    c2<-row.names(contrast)[contrast[,i] == -1 ] ## names of the second conditions of the pairwise comparison
    
    samplesc1 <- row.names(design)[row(as.matrix(design[,c1]))[design[,c1]==1]] ## names of the samples belonging to the first conditions of the pairwise comparison
    samplesc2 <- row.names(design)[row(as.matrix(design[,c2]))[design[,c2]==1]] ## names of the samples belonging to the second conditions of the pairwise comparison
    
    colc1 <- which(colnames(norm.data) %in% samplesc1) ## columns number in the dataset of the samples for the first comparison of the pairwise comparison
    colc2 <- which(colnames(norm.data) %in% samplesc2) ## columns number in the dataset of the samples for the second comparison of the pairwise comparison
    
    p.value<-apply(norm.data,1,function(x){t.test(as.numeric(x[colc2]),as.numeric(x[colc1]), alternative = "t") $p.value}) ## calcul of the p.value of the student test between the conditions of the pairwise comparison
    adjust.p.value<-p.adjust(p.value, method = "BH") ## corrrection of the p.value with the Benjamini & Hochberg technique
    
    fc<-rowMeans(norm.data[,colc1])-rowMeans(norm.data[,colc2]) ## fold change calcul between the conditions of the pairwise comparison
    
    ## stocking the data frame containing the results in a list :
    Results[[i]]<-data.frame(protein=c(row.names(norm.data)))
    Results[[i]][paste("p.value.",colnames(contrast)[i],sep="")]=c(p.value)
    Results[[i]][paste("adjust.p.value.",colnames(contrast)[i],sep="")]=c(adjust.p.value)
    Results[[i]][paste("fc.",colnames(contrast)[i],sep="")]=c(fc)
    print(Results[[i]])
  }

```

##4.3 Differentially expressed proteins

A protein is considered differentially expressed if its p.value and fold change are superior to a determined threshold. The fold change represent how much a protein is differentially expressed between two conditions and the p value allows to evaluate the statistical significance of this difference.

```{r}

thresh_fc <- 0.5
thresh_p <- 0.05
deProt <- Results
for(i in 1:length(Results)){
    fc = as.data.frame(deProt[[i]])[,4]
    p = as.data.frame(deProt[[i]])[,3]
    dt <-as.data.frame(deProt[[i]])
    deProt[[i]] <- dt[which(p<=thresh_p & abs(fc)>=thresh_fc),]
    print(as.data.frame(deProt[[i]]))
  }
```


##4.4 Volcano Plot

For each pairwise comparison in the results list a volcano plot is created, volcano plots are the best way to visualise the p value and the fold change at the same time.

```{r}

library(ggplot2)
library(ggrepel)

tresh_fc = 0.5 ## Fold change threshold
tresh_p = 0.05 ## p.value treshold

for(i in 1:length(Results)){
  plotTitle <- substr(colnames(Results[[i]])[2], 9 ,nchar(colnames(Results[[i]])[2]))
  values <- as.data.frame(Results[[i]])
  forplot <- data.frame(x=as.numeric(values[,4]), y=-log10(values[,3]), id=as.character(values[,1]))
  tmp <- forplot[as.numeric(forplot$y)>=-log10(tresh_fc) & abs(forplot$x)>tresh_fc,]
  p <- ggplot(forplot) + geom_point(aes(x, y , color = ifelse(y>=-log10(tresh_p) & abs(x)>=tresh_fc, "not signi", "FC")),show.legend = F) +
    scale_color_manual(values = c("blue", "red")) +
    geom_text_repel(data = subset(forplot, abs(forplot$x)>=tresh_fc & forplot$y>=-log10(tresh_p)),
                    aes(x,y,label = id),
                    size = 2) +
    geom_vline(xintercept = tresh_fc ) + 
    geom_vline(xintercept = -tresh_fc) + 
    geom_hline(yintercept = -log10(tresh_p)) + 
    labs(title = plotTitle,x="log2(Fold-change)", y="-log10(P.Value)") + theme_bw() 
    print(p)
}
```



<!--chapter:end:SWATH-scl-student.Rmd-->

